platform/jewel/int-ui/int-ui-standalone/src/main/kotlin/org/jetbrains/jewel/intui/standalone/component/JDialogRenderer.kt

```diff
@@ -0,0 +1,362 @@
+                }
+            },
+        measurePolicy = { _, _ -> layout(0, 0) {} },
+    )
+
+    val currentOnDismissRequest = rememberUpdatedState(onDismissRequest)
+    val currentOnKeyEvent = rememberUpdatedState(onKeyEvent)
+
+    val overriddenOnKeyEvent =
+        if (properties.dismissOnBackPress && onDismissRequest != null) {
+            // No need to remember this lambda, as it doesn't capture any values that can change.
+            { event: KeyEvent ->
+                val consumed = currentOnKeyEvent.value?.invoke(event) ?: false
+                if (!consumed && event.isDismissRequest()) {
+                    currentOnDismissRequest.value?.invoke()
+                    true
+                } else {
+                    consumed
+                }
+            }
+        } else {
+            onKeyEvent
+        }
+
+    val composePanel = remember {
+        ComposePanel().apply {
+            layout = null
+            isOpaque = false
+            background = java.awt.Color(0, 0, 0, 0)
+
+            setContent {
+                ProvideValuesFromOtherContext(compositionLocalContext) {
+                    val positionProvider = currentPopupPositionProvider.value
+                    val parentBounds = parentBoundsInRoot.value ?: return@ProvideValuesFromOtherContext
+
+                    println("parentBounds: $parentBounds")
+
+                    Layout(
+                        content = {
+                            currentContent.value()
+
+                            val focusManager = LocalFocusManager.current
+                            LaunchedEffect(Unit) {
+                                if (properties.focusable) {
+                                    focusManager.moveFocus(FocusDirection.Enter)
+                                }
+                            }
+                        },
+                        modifier =
+                            Modifier.focusable(properties.focusable)
+                                .semantics { popup() }
+                                .thenIf(onPreviewKeyEvent != null) {
+                                    Modifier.onPreviewKeyEvent(onPreviewKeyEvent ?: { false })
+                                }
+                                .thenIf(overriddenOnKeyEvent != null) {
+                                    Modifier.onKeyEvent(overriddenOnKeyEvent ?: { false })
+                                },
+                        measurePolicy =
+                            remember(positionProvider, parentBounds) {
+                                JPopupMeasurePolicy(window, popupPositionProvider, parentBounds) { position, size ->
+                                    popupRectangle.value = Rectangle(position.x, position.y, size.width, size.height)
+                                }
+                            },
+                    )
+                }
+            }
+        }
+    }
+
+    val dialog = remember {
+        JDialog(window).apply {
```

Comment:
Any reason why we use JDialog and not JPopup? Does JDialog work with "light" popups like tooltips?

Reply:
So, I haven't found any JPopup; just the JPopupMenu and the Popup from Swing.

* The JPopupMenu is used to build a list of options in a menu format. We can't properly add a compose panel as content to render it (at least I haven't found a way for it);
* The Swing Popup version only exposes methods to show/hide the content. So we would need to already have the position and size of it (which does not happen until the first render);

Also, when I checked the [CfD implementation of 'native windows'](https://github.com/JetBrains/compose-multiplatform-core/blob/2d77505b41717cd8474a5b2751de880f3827104c/compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/WindowComposeSceneLayer.desktop.kt#L65-L78), they are using the JDialog class.

Reply:
Just to keep it documented here too, I also tested using the `DialogWindow` composable, but it has the same issues as the official implementation of native windows (using the 'compose.layers.type' flag).